libs/core/executors/include/hpx/executors/parallel_scheduler.hpp

```diff
@@ -0,0 +1,285 @@
+                        // std::cout << "Task executing on pool" << std::endl;
+                        hpx::execution::experimental::set_value(r);
+                    });
+                },
+                [&](std::exception_ptr ep) {
+                    // std::cerr << "Error occurred" << std::endl;
+                    hpx::execution::experimental::set_error(
+                        HPX_MOVE(os.receiver), HPX_MOVE(ep));
+                });
+        }
+    };
+
+    template <typename Shape, typename F>
+    struct bulk_sender
+    {
+        parallel_scheduler scheduler;
+        Shape shape;
+        F f;
+
+        bulk_sender(parallel_scheduler&& sched, Shape sh, F&& func)
+          : scheduler(HPX_MOVE(sched))
+          , shape(sh)
+          , f(HPX_FORWARD(F, func))
+        {
+            // std::cout << "Bulk sender created with shape: " << shape << std::endl;
+        }
+
+#if defined(HPX_HAVE_STDEXEC)
+        using sender_concept = hpx::execution::experimental::sender_t;
+#endif
+        using completion_signatures =
+            hpx::execution::experimental::completion_signatures<
+                hpx::execution::experimental::set_value_t(),
+                hpx::execution::experimental::set_error_t(std::exception_ptr),
+                hpx::execution::experimental::set_stopped_t()>;
+
+        template <typename Env>
+        friend auto tag_invoke(
+            hpx::execution::experimental::get_completion_signatures_t,
+            bulk_sender const&, Env) noexcept -> completion_signatures
+        {
+            return {};
+        }
+    };
+
+    template <typename Receiver, typename Shape, typename F>
+    struct bulk_operation_state
+    {
+        parallel_scheduler scheduler;
+        Receiver& receiver;    // Store by reference
+        Shape shape;
+        F f;
+        std::shared_ptr<std::atomic<int>> tasks_remaining;
+
+        bulk_operation_state(
+            parallel_scheduler&& sched, Receiver& r, Shape sh, F&& func)
+          : scheduler(HPX_MOVE(sched))
+          , receiver(r)
+          , shape(sh)
+          , f(HPX_FORWARD(F, func))
+          , tasks_remaining(
+                std::make_shared<std::atomic<int>>(static_cast<int>(shape)))
+        {
+            // std::cout << "Bulk operation state created" << std::endl;
+        }
+
+        friend void tag_invoke(start_t, bulk_operation_state& os) noexcept
+        {
+            // std::cout << "Bulk start() called" << std::endl;
+            hpx::detail::try_catch_exception_ptr(
+                [&]() {
+                    thread_pool_scheduler exec{os.scheduler.get_thread_pool()};
+                    for (Shape i = 0; i < os.shape; ++i)
+                    {
+                        exec.execute([i, &os]() mutable {
+                            // std::cout << "Bulk task executing for index: " << i <<;
+                            os.f(i);
+                            if (--(*os.tasks_remaining) == 0)
```

Comment:
Using `std::atomic_ref` instead of a `std::shared_ptr` would be preferrable (at least for C++20)